### ISSUE TITLE
Round 10 peak Dropwizard results

### ISSUE THREAD

**Issue by zloster** [TechEmpower/FrameworkBenchmarks#1545]
Dropwizard and dropwizard-mongodb don't have results from `peak` environment for round 10. The error in the log files is:
```
Error: Port 9000 is not available, cannot start dropwizard
```
This doesn't seem right.

The log files are here:
* https://github.com/TechEmpower/TFB-Round-10/blob/master/peak/linux/results-2015-03-24-peak-final/latest/logs/dropwizard-mongodb/err.txt 
* https://github.com/TechEmpower/TFB-Round-10/blob/master/peak/linux/results-2015-03-24-peak-final/latest/logs/dropwizard/err.txt

**Comment by bhauer** [comment]
This was mentioned on the Google Group as well:

https://groups.google.com/d/msg/framework-benchmarks/7wU8X15C3yE/0wP1dcmkGXgJ

I believe there are two issues that we must deal with, both of which are related to the toolset and not specifically to DropWizard:

* The aggressive culling of processes that we have is still not aggressive enough.  We are allowing processes to start under alternative usernames and not killing those when the frameworks' `stop` scripts fail to do so themselves.  So a previous framework lingered into the DropWizard test window preventing DropWizard from being able to start.  I believe we have a separate task to prevent the ability to start processes as other users to counter this.

* DropWizard did not start because of the port being unavailable.  This is actually a *graceful* failure where we have to-date mostly seen less graceful failures.  In fact, the failure was so graceful that our toolset didn't log it as a failure at all.  As a result, DropWizard isn't captured in the list of failed tests on the results web site.  We should either capture did-not-start scenarios such as this as failures for the purposes of rendering the results *or* capture them as a new kind of event ("did-not-start" for example) and make those available on the results site.

**Comment by bhauer** [comment]
This issue is probably related to issue #1553.

**Comment by shkhln** [comment]
>  So a previous framework lingered into the DropWizard test window preventing DropWizard from being able to start.

Is it so? As far is I can tell, the benchmark runs through the framework tests in alphabetical order and not a single test before dropwizard is configured to listen on port 9000.

**Comment by yunspace** [comment]
As a quick work around I can change both dropwizard and dropwizard-mongodb to use different ports.

**Comment by richdougherty** [comment]
FYI on the [related Play issue](#1553) I [said](https://github.com/TechEmpower/FrameworkBenchmarks/issues/1553#issuecomment-98881180) that I'd prefer to fix the underlying test cleanup issue rather than change the port that tests bind to. I think it's good if tests bind to the same port because that stops tests running at the same time and potentially messing up benchmark results.

**Comment by hamiltont** [comment]
Weighin in on why this is not currently an easy fix - [here](https://travis-ci.org/TechEmpower/FrameworkBenchmarks/builds/60018192) is this list of how much would break if we fixed the underlying issue properly by removing sudo permissions from the testrunner user. I'm totally in favor of fixing it properly - just pointing out that doing so at this time would break *a lot*. 

Granted, it's quite sad that some people who worked so hard with us between R9/10 to make their implementations solid have been shafted by previously-running frameworks that we could not protect against. So for the moment, I'm in favor of fixing this manually as much as possible, e.g. hard-stop the entire benchmark if the port is bound and inform the operator that a manual cleanup (or reboot, whatever is easier) is needed. This would **really** increase the load on @msmith-techempower during a round (e.g. maybe two previews, any more would be ridiculous), but the only other fixes I see are heuristic e.g. locate/SIGKILL the one process holding the port (but no guarantee it's workers or children die), locate/SIGKILL the session of that process (again, no guarantee that *all* workers/children die), etc. 

So rock and a hard place it seems, we either make it 10x harder to run a round, or break 30-50% of the existing frameworks.

**Comment by shkhln** [comment]
Again, which process *even had the opportunity* to bound to port 9000 before dropwizard? How about logging `sudo netstat -pltn` output before each test? It won't break anything, I promise :)

**Comment by hamiltont** [comment]
> Again, which process even had the opportunity to bound the port 9000 before Dropwizard?

[here](https://raw.githubusercontent.com/TechEmpower/TFB-Round-10/master/peak/linux/results-2015-03-24-peak-final/2015-03-24.txt) is the full run log, searching for dropwizard shows that we were at ~21% progress when that test started, so ~20% of the [total of 293 configurations](https://www.techempower.com/blog/2015/04/21/framework-benchmarks-round-10/) ran before dropwizard. Also, note that it's common for the benchmark_config to only list the port of the load balancer, while internally many workers spin up on arbitrary upstream ports (e.g. take a look as how aspnet starts up). 

I suppose you could go thru the entire run log, find everything listed before dropwizard, and examine it's setup process to identify the candidates that might have locked down the port, but you're right that adding logging is the much better fix

> How about logging sudo netstat -pltn output before starting each test? It won't break anything, I promise :)

Great idea. Any other "debug" commands anyone can recommend?

**Comment by hamiltont** [comment]
@msmith-techempower - given that multiple frameworks had this "port not available" issue, meganuke is getting circumvented by testrunner's sudo perms. Can you add the logging stuff that we originally had (like netstat) back in before R11? I'd vote to make the logging in this case ridiculously verbose, as it's not a terribly common thing (e.g. only these two reported cases that I know of). I would seriously should consider making the "port not available" actually terminate the current round and tell the operator to manually fix the issue

**Comment by msmith-techempower** [comment]
I'm fine with bumping the logging up a lot on this issue, but terminating the round is probably overkill.

**Comment by yunspace** [comment]
I confirm @shkhln observation in that based on the configurations been ran in alphabetic order, there are not port 9000 before dropwizard. However as @hamiltont pointed out, there may be internal ports which are not listed in the config. I think the logging would definitely help. 

```
./frameworks/Java/dropwizard/benchmark_config.json:      "port": 9000,
./frameworks/Java/dropwizard-mongodb/benchmark_config.json:      "port": 9000,
./frameworks/Java/play2-java/benchmark_config.json:        "port": "9000", 
./frameworks/Java/play2-java/benchmark_config.json:        "port": "9000", 
./frameworks/Java/play2-java/benchmark_config.json:        "port": "9000", 
./frameworks/Java/play2-java/benchmark_config.json:        "port": "9000", 
./frameworks/Java/play2-java/benchmark_config.json:        "port": "9000", 
./frameworks/Scala/akka-http/benchmark_config.json:        "port": 9000,
./frameworks/Scala/play2-scala/benchmark_config.json:        "port": "9000", 
./frameworks/Scala/play2-scala/benchmark_config.json:        "port": "9000", 
./frameworks/Scala/play2-scala/benchmark_config.json:        "port": "9000", 
./frameworks/Scala/play2-scala/benchmark_config.json:        "port": "9000", 
./frameworks/Scala/play2-scala/benchmark_config.json:        "port": "9000", 
./frameworks/Scala/play2-scala/benchmark_config.json:        "port": "9000", 
./frameworks/Scala/unfiltered/benchmark_config.json:      "port": 9000,
```
